### ISSUE TITLE
Crucible fails the live repair test

### ISSUE THREAD

**Issue by leftwo** [oxidecomputer/crucible#893]
This dependabot PR, chore(deps): update rust crate tokio to 1.32 #890, failed the live_repair test:

https://github.com/oxidecomputer/crucible/pull/890/checks?check_run_id=16220798170

We need to figure out why.  Nothing in that PR should have any impact on live repair, which suggests
there is an underlying issue with live repair that we have not seen yet.  Either that, or something in
Tokio 1.32 broke in a strange way to introduce this.

**Comment by leftwo** [comment]
Output printed for the failing test:
[test-repair-fail-testout.log](https://github.com/oxidecomputer/crucible/files/12444133/test-repair-fail-testout.log)

The crutest log (not including stedrr, an oversight that will get fixed, stderr is in the above log)
[crutest-replace.log](https://github.com/oxidecomputer/crucible/files/12444135/crutest-replace.log)

The log from DSC (the downtstairs)
[dsc.log](https://github.com/oxidecomputer/crucible/files/12444136/dsc.log)

**Comment by leftwo** [comment]
The test was running this command:
```
ptime -m /var/tmp/bins/crutest replace -t 127.0.0.1:8810 -t 127.0.0.1:8820 -t 127.0.0.1:8830 --replacement 127.0.0.1:8840 -g 1 -c 10 --stable
```

The panic/mismatch is:
```
2023-08-25T17:44:24.943Z thread 'tokio-runtime-worker' panicked at '[2] read hash mismatch on 4993 upstairs/src/lib.rs:4438:29                                                                                                                  
2023-08-25T17:44:24.950Z Expected [Some(c05cf20bce896a22), Some(c69f8cfabdfa715f), Some(14b1d79450301735)]                                                                                                                                      
2023-08-25T17:44:24.959Z Computed [Some(4b685ce4967184af), Some(c69f8cfabdfa715f), Some(14b1d79450301735)]
``` 
By default (which this test uses), `dsc` will create 4k blocks, 100 blocks per extent, and 15 extents.
The downstairs are not running with encryption.

The Job that had the mismatch was coming  from downstairs 2 (from the panic message)
Here is the job itself:  Note that sometimes the hash is printed in hex, sometimes decimal, but the
decimal values are the same as the hex.
```
job: DownstairsIO {
    ds_id: 4993,
    guest_id: 3994,
    work: Read {
        dependencies: [],
        requests: [
            ReadRequest { eid: 12, offset: Block { value: 8, shift: 12 } },
            ReadRequest { eid: 12, offset: Block { value: 9, shift: 12 } },
            ReadRequest { eid: 12, offset: Block { value: 10, shift: 12 } }
        ]
    },
    state: {2: Done, 1: Skipped, 0: Done},
    ack_status: AckReady,
    replay: false,
    data: Some([
        ReadResponse {
            eid: 12,
            offset: Block { value: 8, shift: 12 },
            data: b"\xbc\x06\x06...\x06\x06\x06\x",
            block_contexts: [BlockContext { hash: 13861219885663218210, encryption_context: None }]
        },
        ReadResponse {
            eid: 12,
            offset: Block { value: 9, shift: 12 },
            data: b"\xbd\x04\x04...\x04\x04\x04",
            block_contexts: [BlockContext { hash: 14312313149363745119, encryption_context: None }]
        },
        ReadResponse {
            eid: 12,
            offset: Block { value: 10, shift: 12 },
            data: b"\xbe\x04\x04...\x04\x04\x04",
            block_contexts: [BlockContext { hash: 1491209983636805429, encryption_context: None }]
        }
    ]),
    read_response_hashes: [
        Some(13861219885663218210),
        Some(14312313149363745119),
        Some(1491209983636805429)
    ],
    impacted_blocks: InclusiveRange(ImpactedAddr { extent_id: 12, block: 8 }, ImpactedAddr { extent_id: 12, block: 10 })
}
```

**Comment by leftwo** [comment]
From the job,
We know its a read of three blocks coming back from downstairs 2.
Downstairs 1 job was `skipped`
Which means, the data in the hash was probably populated by a read from downstairs 0.

The `DownstairsIO` job output from above tells us that extent 12 block 8 is where we have the mismatch.
The other two blocks (9, 10) have matching hashes.

Since encryption is off, we can see what data was in the block.
To figure out the physical block offset, we have `eid: 12` and block `value: 8`.
 block 12 * 100 (each extent has 100 blocks in it) 1200 + 8 = 1208
To get the value that crutest will put at the start of the block, it's block number % 255, 1208 % 255 = 188 or 0xbc
So that all checks out.
The data in the block is all 0x06, which tells us on this downstairs (2) crutest has written to block 1208 six times.
(Technically at least 6 times, the counter will roll over at 255, but we don't have enough IO going
on with this job to roll over the IO counter)

Given this info, block 1208, six expected writes, we can go back through out IO log output from the test
and see if/when those six writes happened, and possibly find more clues.
We don't have the actual downstairs region files any longer, so we don't know what we were expecting
to see, though it should be possible to re-create buffer contents with write counters around six and hash 
that buffer and see what has matches.

**Comment by leftwo** [comment]
Some notes after looking through the IO log from the test output

Here is the write to block 1208 where we have written the six:
```
                          Block             12  12  12  12
                          Number            08  09  10  11
                                            --------------
164/800 Write block 1208  len 28672  data:   6   4   4   5   5   5   7
                                             ^
                                             |
                                        Here is our 6.
```
Messages around the same time, showing we are repairing extent 11 when the write lands.
```
{"msg":"Write to Extent 11:98:12 under repair","v":0,"name":"crucible","level":40,"time":"2023-08-25T17:44:18.997682887Z","hostname":"ip-10-150-1-115.us-west-2.compute.internal","pid":1161}
{"msg":"IO Write 4726 on eur 11 Added deps 12","v":0,"name":"crucible","level":40,"time":"2023-08-25T17:44:18.997872783Z","hostname":"ip-10-150-1-115.us-west-2.compute.internal","pid":1161}
```

Here is the write that we have "lost", coming later in the log
```
                                             12  12  12  12  12  12  12  12
                                             01  02  03  04  05  06  07  08
                                             ------------------------------
678/800 Write block 1201  len 32768  data:    9  10   9   7   8   9   6   7
                                                                          ^
                                                                          |
                                                                 Here is the 7 that
                                                                 I suspect has the 
                                                                 matching hash
```
Repair happens when IO is in progress.
0 is offline when we do the write 164/800, the 6 to 1208, that we see on the read.
Repair of extent 12 happens just after we see issuance of the lost
write and then a flush output in the log.  As the log is printed before we
submit the write to the upstairs, we can't tell if the actual write lands
before the repair starts, or after.

Here is the extent repair of 12, source 2, destination, 0.
```
{"msg":"Repair for extent 12 s:2 d:[0]","v":0,"name":"crucible","level":30,"time":"2023-08-25T17:44:19.089775674Z","hostname":"ip-10-150-1-115.us-west-2.compute.internal","pid":1161,"":"downstairs"}
```

The read happens much later, so it's not possible that we have an old Read or
a replay read that could mismatch.

**Comment by leftwo** [comment]
I'm able to duplicate this outside of CI, and have done so now several times.
I find that running debug bits was required, I ran over 80,000 Live Repairs and did not see a problem.
It was not till I switched to debug bits (like the test has) that I reproduced the failure.

You can modify the `main_port` to a unique number and run this script multiple times on the
same system.

```
#!/bin/bash
#
# Attempt to reproduce the live-repair test failure that we saw
# in CI.
export BINDIR=./target/debug

# starting_count==26310 (on atrium)
# On gimlet-sn05, 1 got to 36,600 with release build
#
starting_count=0
loop=0

main_port=9900
dsc_log="/tmp/dsc-${main_port}.log"
dsc_root="/var/tmp/dsc/${main_port}"
test_log="/tmp/crutest-replace-${main_port}.log"

(( port_base = main_port + 10))

(( ds0_port = port_base))
(( ds1_port = ds0_port + 10))
(( ds2_port = ds1_port + 10))
(( ds3_port = ds2_port + 10))

# Repairs per crutest loop
rpc=30

while :; do
    ((loop += 1))
    echo "[${main_port}] Create dsc"
    $BINDIR/dsc create \
      --ds-bin "$BINDIR"/crucible-downstairs \
      --region-count 4 \
      --region-dir ${dsc_root}/dsc/regions \
      --output-dir ${dsc_root}/dsc/output \
      --port-base ${port_base} \
      --cleanup > ${dsc_log} 2>&1
    if [[ $? -ne 0 ]]; then
        echo "[${main_port}] create dsc failed at $loop"
        break
    fi

    echo "[${main_port}] start dsc"
    $BINDIR/dsc start \
      --ds-bin "$BINDIR"/crucible-downstairs \
      --region-dir ${dsc_root}/dsc/regions \
      --output-dir ${dsc_root}/dsc/output \
      --port-base ${port_base} \
      --control 127.0.0.1:${main_port} \
      --region-count 4 >> ${dsc_log} 2>&1 &
    dsc_pid=$!
    sleep 5

    if [[ "$dsc_pid" -eq 0 ]]; then
        echo "[${main_port}] dsc_pid is zero, which is bad, exit"
        cat ${dsc_log} || true
        exit 1
    fi

    if ps -p "$dsc_pid" > /dev/null; then
        echo "[${main_port}] Found dsc running, ${dsc_pid}  Run crutest replace"
    else
        echo "[${main_port}] dsc failed"
        cat ${dsc_log} || true
        exit 1
    fi

    "$BINDIR"/crutest replace \
      -t 127.0.0.1:${ds0_port} \
      -t 127.0.0.1:${ds1_port} \
      -t 127.0.0.1:${ds2_port} \
      --replacement 127.0.0.1:${ds3_port} \
      --verify-at-end \
      -g 1 -c ${rpc} --stable > ${test_log} 2>&1
    if [[ $? -ne 0 ]]; then
        echo "[${main_port}] Test failed at $loop"
        echo "dsc root at: ${dsc_root}"
        break
    fi

    echo "[${main_port}] Stop DSC ${dsc_pid}, wait"
    $BINDIR/dsc cmd --server http://127.0.0.1:${main_port} shutdown
    wait "$dsc_pid"
    (( total = loop * rpc ))
    (( total = total + starting_count ))
    echo "[${main_port}] $(date) at $loop, total: $total"
done

echo "[${main_port}] Test ends on $(date) with $loop"
$BINDIR/dsc cmd --server http://127.0.0.1:${main_port} shutdown 2> /dev/null
```

**Comment by leftwo** [comment]
I saw this panic again on another system:
```
thread 'tokio-runtime-worker' panicked at '[0] read hash mismatch on id 2529
                                                                                                               
Expected [Some(562750943e4b2e78), Some(4b58c84d94aca2ac), Some(d90ceb6856901b2b), Some(8620ec8d3e9e080f), Some(6f45de8b86016e44), Some(fa08071078ad5e53), Some(9a906caf57a4a5c0)]
Computed [Some(4876bf4e7945c564), Some(26ab5338eaff76d6), Some(f68789d807fd0f2e), Some(b3d6ba7a60ba51dc), Some(6f45de8b86016e44), Some(fa08071078ad5e53), Some(9a906caf57a4a5c0)] 
             
guest_id:1530                                                                                                                                                                              
request:[
    ReadRequest { eid: 6, offset: Block { value: 69, shift: 12 } },
    ReadRequest { eid: 6, offset: Block { value: 70, shift: 12 } },
    ReadRequest { eid: 6, offset: Block { value: 71, shift: 12 } },
    ReadRequest { eid: 6, offset: Block { value: 72, shift: 12 } },
    ReadRequest { eid: 6, offset: Block { value: 73, shift: 12 } },
    ReadRequest { eid: 6, offset: Block { value: 74, shift: 12 } },
    ReadRequest { eid: 6, offset: Block { value: 75, shift: 12 } }
]
job state:{2: Done, 0: Done, 1: Done}',
```

From the above, (and knowing we have 100 blocks per exetnt), we have hash mismatch on blocks 669, 670, 671, 672.

Sure enough, if we look at the actual downstairs region files, we can see the mismatch even though the
flush/gen/dirty are all consistent:

```
/crucible-downstairs dump -d /stock-0/dsc/11000/dsc/regions/11020 -d /stock-0/dsc/11000/dsc/regions/11030 -d /stock-0/dsc/11000/dsc/regions/11040 -e 6

    Extent 6                                                                                                                                                                                                     
GEN             1        1        1    
FLUSH_ID      145      145      145                          
DIRTY                                                                                                                                                                                                                                    
BLOCK  D0 D1 D2  C0 C1 C2 DIFF                                                                                                                                                                                           
...
  668   A  A  A   A  A  A
  669   A  A  C   A  A  C <---
  670   A  A  C   A  A  C <---
  671   A  A  C   A  A  C <--- 
  672   A  A  C   A  A  C <--- 
  673   A  A  A   A  A  A                                                                                                                                                                                                 
  674   A  A  A   A  A  A
...
```

From the log for this test  I can explain it all:
We have downstairs 0 under repair.
We have started extent 5 repairing.

Here is the log describing the jobs that will repair extent 5:
```
RE:5 repair extent with ids 2220,2221,2222,2223 deps:[2219] 
```
These repair jobs are now in flight operating on extent 5.

A write (id 2234) comes in that spans extents 5 and 6:
```
1104/1800 Write block  592  len 36864  data:   5   4   5   5   3   2   3   3   2                                                                                                                                                      
Write to Extent:5 Block:92 under repair                                                                                                                                                           
Write to Extent:5 Block:93 under repair                                                                                                                                                                                 
Write to Extent:5 Block:94 under repair                                                                                                                                                                                 
Write to Extent:5 Block:95 under repair                                                                                                                                                                                 
Write to Extent:5 Block:96 under repair                                                                                                                                                                                 
Write to Extent:5 Block:97 under repair                                                                                                                                                                                 
Write to Extent:5 Block:98 under repair                                                                                                                                                                                 
Write to Extent:5 Block:99 under repair                                                                                                                                                                                 
Write to Extent:6 Block:0 past extent under repair 5                                                                                                                                                                    
IO Write 2234 on eur 5 Added deps 6                                                                                                                                                                                     
IO Write 2234 has deps [2223, 2221, 2220, 2230, 2231, 2232, 2233]                                                                                                                                                       
[0] 2234 final dependency list [2223, 2221, 2220, 2230, 2231, 2232, 2233]
```

Because we are repairing this extent, and this IO spans on to the next extent, we go off 
and create future repair job IDs 2230, 2231, 2231, and 2233 for the repair of extent 6.  
These all get put on IO 2234's dependency list.                                                                                                                                               

Then, a few other IOs come along, and eventually a flush.  This flush (ID 2240) gets all the
outstanding IOs as dependencies, including the write job ID 2234 that is the extent spanning write 
but not the future repair work.
Here is the flush IOs dependency list:
```                                                                    
IO Flush 2240 has deps [2239, 2238, 2237, 2236, 2235, 2234, 2229, 2228, 2227, 2226, 2225, 2224, 2223, 2221, 2220]                                                                                                       
```

Now, we send flush id 2240 to the downstairs, but at the time we send it, the `extent_limit` is 5,
we include this `extent_limit` in the flush when we send it to the downstairs.
However, this flush depends on job ID 2234 which depens on the future repair work ids: 2230-2233.

What the `under repair` downstairs will do is flush extents that are at or below the `extent_limit` it
receives from the upstairs.  This is a different bug, and not actually the problem here, but what the 
flush should do is flush through extent 6, as this flush is dependent on a job that depends on repair 
for extent 6.

Anyways, back to the problem we actually see in this situation.

Another write shows up, this will be the "missing write"
```
1111/1800 Write block  669  len 16384  data:   5   4   4   4                                                                                                                                                          
IO Write 2242 has deps [2240]
```
So, this write for extent 6 will wait for the flush (2240) to finish.
We send this write off to the downstairs, but here is the problem.
During `enqueue()`, we look to see if an IO is:
below, at, or above the extent under repair.                                                                                                                                                    

If it is below, we send it to all three downstairs.
If it is at, we send it to all three downstairs (because the repair is in progress and we will
come after the repair, so it should all be good.
If the IO is above the extent under repair, we `skip` that IO on a downstairs we 
have not repaired yet.  We do this because we don't know what is on that downstairs yet,
and when it gets repaired, we will include this IO in that repair.

So, in our situation above, we have our IO 2242 that depends on other IOs that depend
on the future repair of extent 6, but we decided (in `enqueue()`) that 2242 should be
skipped on the downstairs under repair.

We end up with the write not making it to that downstairs.
Then, later, another flush will come along and we believe this downstairs is the same as
the other two and we update the flush/gen numbers, clear the dirty bit, and move on.

This leaves us with a mismatch waiting to happen.
Eventually, a read touches these blocks and we panic.

**The TL;DR summary to re-create this situation is:**
LiveRepair is in progress
An extent is being repaired, for example, say extent 5
A write arrives that spans extent 5 and 6 that leads to creation of future job IDs.
A flush comes after the write that depends on the write. 
A write for extent 6 arrives (not overlapping with the first write) that depends on the flush.
The repair of extent 5 finishes.

**Comment by leftwo** [comment]
I believe a fix for both the issues here could be, when we enqueue an IO, we not only
look for what extent (if any) is being repaired, but also check for future repair reservations.
If any future repair reservations exist, they must be taken into account for proper dependency
tracking as well as what the extent limit will be for a flush.

**Comment by leftwo** [comment]
I don't have the actual data from the run above, but I do have several other enample
 mismatches where the mismatch data had a write counter that was one less
than the non-mismatch data.   To be more specific, a crutest write will write the
block number (%255), then a counter of the number of times crutest has written
to the block in question.  In the three or four other example panics I have from
this, the "mismatch" write counter has always been one less than the "good"
write counter.

This "one less" fits in the overall pattern where it looks like one downstairs "loses"
a write.
